server/src/main/java/org/elasticsearch/cluster/metadata/MetadataMappingService.java

```diff
@@ -360,4 +383,48 @@ public TimeValue ackTimeout() {
                     }
                 });
     }
+
+    private void populateColumnPositions(Map<String, Object> mapping, Integer maxPosition, CompressedXContent mappingToReference) {
+        Map<String, Object> parsedTemplateMapping = XContentHelper.convertToMap(mappingToReference.compressedReference(), true).map();
+
+        populateColumnPositionsImpl(
+            Maps.getOrDefault(mapping, "default", mapping),
+            new Integer[]{maxPosition},
+            Maps.getOrDefault(parsedTemplateMapping, "default", parsedTemplateMapping)
+        );
+    }
+
+    private void populateColumnPositionsImpl(Map<String, Object> indexMapping, Integer[] columnPosition, Map<String, Object> templateMapping) {
+        Map<String, Object> indexProperties = Maps.get(indexMapping, "properties");
+        if (indexProperties == null) {
+            return;
+        }
+        Map<String, Object> templateProperties = Maps.get(templateMapping, "properties");
+        if (templateProperties == null) {
+            templateProperties = Map.of();
+        }
+        for (var e : indexProperties.entrySet()) {
+            String key = e.getKey();
+            Map<String, Object> indexColumnProperties = (Map<String, Object>) e.getValue();
+            Map<String, Object> templateColumnProperties = (Map<String, Object>) templateProperties.get(key);
+
+            if (templateColumnProperties == null) {
+                templateColumnProperties = Map.of();
+            }
+            templateColumnProperties = furtherColumnProperties(templateColumnProperties);
+            indexColumnProperties = furtherColumnProperties(indexColumnProperties);
+
+            Integer indexChildPosition = (Integer) indexColumnProperties.get("position");
+            if (indexChildPosition == null) {
+                Integer templateChildPosition = (Integer) templateColumnProperties.get("position");
+                if (templateChildPosition != null) {
+                    indexColumnProperties.put("position", templateChildPosition);
+                } else {
+                    indexColumnProperties.put("position", ++columnPosition[0]);
+                }
+            }
```

Comment:
would this have to add the position to both the template and the index?

Reply:
I have looked into `update`/`insert`/`add column` and each uses a version of `updateTemplate` methods to update the template mappings (if necessary). So by this point, templates should already have the latest mapping.